README
tibrv module for Qore
version 0.1 by David Nichols

The "tibrv" module provides direct TIBCO Rendezvous(R) support, allowing qore programs to send and receive Rendezvous messages directly.

Requires qore 0.7.0 and TIBCO Rendezvous(R) headers and libraries to build.
Note that the TIBCO C++ library (libtibrvcpp.a or libtibrvcpp64.a) must be built with the same C++ compiler that is used to build the qore library and this module.

TIBCO Rendezvous(R) is a registered trademark of TIBCO, Inc.

Documentation in docs/tibrv-module-doc.html
(built from docbook sources in docs/docbook)

test/example script: test/tibrv-test.q

To use TIBCO Rendezvous support use the "%use tibrv" directive to ensure that the functionality is loaded
(see test/tibrv-test.q for more info)

Classes supplied by this module:
 + TibrvListener         - for listening to reliable messages
 + TibrvSender           - for sending reliable messages
 + TibrvCmListener       - for listening to certified messages
 + TibrvCmSender         - for sending certified messages
 + TibrvFtMember         - to implement a fault-tolerant node
 + TibrvFtMonitor        - to monitor a fault-tolerant group
 + TibrvDistributedQueue - to manage certified messages in a distributed queue

NOTE: we do not send native Rendezvous arrays - we use the old style "multiple fields with the same name" approach to have more flexible arrays (different datatype in each element, etc).   Native Tibrv arrays are decoded properly when received.

automatic type mapping:
  TIBRVMSG_BOOL <-> boolean
  TIBRVMSG_I8,U8,I16,U16,I32,U32,I64,U64 -> int (note there is a loss of precision with the U64 -> int conversion)
  int -> TIBRVMSG_I64
  TIBRVMSG_F32,F64 -> float
  float -> TIBRVMSG_F64
  TIBRVMSG_DATETIME <-> date
  TIBRVMSG_OPAQUE <-> binary
  TIBRVMSG_STRING -> string     (strings are marked with the default encoding for the object)
  string -> TIBRVMSG_STRING     (encoding conversions are made if necessary to the encoding set for the object)
  TIBRVMSG_XML -> string        (to explicitly send this type, use the tibrv_xml() function)
  TIBRVMSG_IPPORT16 -> int      (to explicitly send this type, use the tibrv_ipport16() function)
  TIBRVMSG_IPADDR32 -> string   (produces a string like "192.168.1.1" - to explicitly send this type, use the tibrv_ipaddr32() function)

to do explicit qore->tibrv type mapping, use the following helper functions supplied by the module:
  tibrv_i8()
  tibrv_u8()
  tibrv_i16()
  tibrv_u16()
  tibrv_i32()
  tibrv_u32()
  tibrv_i64()
  tibrv_u64()
  tibrv_f32()
  tibrv_f64()
  tibrv_ipport16()
  tibrv_ipaddr32()  - note: takes a string argument like "192.168.1.1"
  tibrv_xml()
  tibrv_bool()

set certificates for secure daemons using tibrvSetDaemonCert() and tibrvSetUserCertWithKey() using the SSLCertificate and SSLPrivateKey classes.  Note that these functions are only available if your tibrvcpp library was compiled with secure daemon support.

BUILD AND INSTALL
-----------------

Set the RV_ROOT environment variable to the Rendezvous directory before calling configure (or use the --with-tibrv configure option) to build the "tibrv" module for direct Rendezvous support.

Note that to build this module the libtibrvcpp library must be present; on some platforms you have to rebuild this yourself from the sources provided by TIBCO in order for it to link with the C++ compiler you are using - the sources are normally present in $RV_ROOT/src/librvcpp, normally you have to edit the Makefile provided there and then type "make" to rebuild.  I had to include "ranlib libtibrvcpp.a" on the libraries I rebuilt for OS X.  Secure daemon support is turned off by default in tibrvcpp, to enable secure daemon support edit $RV_ROOT/src/librvcpp/Makefile and uncomment the SD_MODULE line near the end of the file, rebuild, install the new library in $RV_ROOT/lib, and rerun qore's configure script.

configure
make
sudo make install

The configure script will find out where your qore module directory is found and set this for the install directory.

BUILD FROM SVN
--------------

reconf.sh
configure
make
sudo make install


ChangeLog
2010-06-21 09:12  david_nichols

	* qore-tibrv-module.spec: updated spec files to generically
	  determine *suse versions for distribution tagging in rpm file
	  names

2010-06-20 06:03  david_nichols

	* ChangeLog: fixed ChangeLogs with non-broken svn2cl

2010-06-14 15:51  david_nichols

	* docs/tibrv-module-doc.html: updated docs

2010-06-14 14:04  david_nichols

	* docs/docbook/get-docbook-root.sh: added missing file

2010-06-10 14:22  david_nichols

	* docs/docbook/Makefile, docs/docbook/tibrv-module.docbook,
	  docs/qore-style.css, docs/tibrv-module-doc.html: updated docs

2010-05-12 11:49  david_nichols

	* configure.ac: solaris and sun studio compile and warning
	  fixes/enhancements

2010-05-11 14:45  david_nichols

	* configure.ac: compile fixes for solaris

2010-05-11 14:44  david_nichols

	* reconf.sh: compile fixes for solaris

2010-05-08 17:32  david_nichols

	* RELEASE-NOTES: updated release notes for new date/time
	  serialization approach; notably that date/time values are send in
	  UTC and assumed to be UTC when received

2010-05-08 11:39  david_nichols

	* src/QoreTibrvTransport.cc, test/tibrv-test.q: fixed typos in
	  nanoseconds field, updated test to include millisecond portion of
	  time for testing

2010-05-08 10:40  david_nichols

	* RELEASE-NOTES, configure.ac, qore-tibrv-module.spec: updated
	  release version to 0.2.2

2010-05-08 10:37  david_nichols

	* src/QoreTibrvTransport.cc: fixed date/time conversions to use new
	  APIs when available; if not available then at least serialize and
	  deserialize date/time values with millisecond resolution

2009-12-09 12:58  david_nichols

	* configure.ac: fixed all modules to use decimal comparisons with
	  the qore version string instead of octal so that they will still
	  compile with qore v0.8.0 and up...

2009-11-22 06:03  david_nichols

	* configure.ac: updated configure to assume 64-bit build if
	  host_cpu = "amd64"

2009-10-30 08:22  david_nichols

	* configure.ac: updated all module configure scripts to use gnutar
	  on darwin to avoid picking up OSX resource files in source
	  releases

2009-10-05 09:09  david_nichols

	* ChangeLog, RELEASE-NOTES, qore-tibrv-module.spec: minor doc
	  updates

2009-09-21 11:07  david_nichols

	* configure.ac: updated module configure scripts to select 64-bit
	  target mode by default on snow leopard i386

2009-09-10 06:49  david_nichols

	* configure.ac: updated configure.ac to properly compile in 32-bit
	  mode on OSX snow leopard (10.6)

2009-07-20 13:16  david_nichols

	* configure.ac, qore-tibrv-module.spec: added opensuse 11.2
	  detection to spec file, fixed rhel detection in spec files,
	  updated configure.ac to work with older versions of autotools

2009-06-08 08:38  david_nichols

	* qore-tibrv-module.spec: fixed qore module dir detection

2009-06-02 05:24  david_nichols

	* qore-tibrv-module.spec: fixed typo in sles version detection

2009-01-08 18:43  david_nichols

	* qore-tibrv-module.spec: simplified spec files a little

2009-01-08 15:13  david_nichols

	* qore-tibrv-module.spec: commented-out additional "Source0" line
	  so that packages will build on Fedora 10 (OBS)

2009-01-08 14:44  david_nichols

	* reconf.sh: updated reconf.sh to look for glibtoolize as well as
	  libtoolize

2009-01-07 13:38  david_nichols

	* RELEASE-NOTES: updated release notes

2009-01-07 13:37  david_nichols

	* Makefile.am, configure.ac, ltmain.sh, qore-tibrv-module.spec,
	  reconf.sh: updated version to 0.2.1; updated reconf.sh to use
	  libtoolize to generate ltmain.sh

2008-11-19 15:19  david_nichols

	* qore-tibrv-module.spec: corrected version number

2008-11-03 15:31  david_nichols

	* ChangeLog, Makefile.am, RELEASE-NOTES, configure.ac,
	  qore-tibrv-module.spec, src/Makefile.am, src/tibrv-module.cc,
	  src/tibrv-module.h: updated to new module handling; updated
	  version to 0.0.2

2008-10-30 17:20  david_nichols

	* qore-tibrv-module.spec: various fixes

2008-10-23 06:30  david_nichols

	* configure.ac: updated to set 64-bit compile mode by default on
	  itanium with g++ and aCC

2008-10-08 20:16  david_nichols

	* qore-tibrv-module.spec, src/Makefile.am: spec file and module
	  build fixes using prefix=/usr and make install
	  DESTDIR=$RPM_BUILD_ROOT

2008-10-07 06:20  david_nichols

	* ChangeLog: updated changelog

2008-09-09 07:19  david_nichols

	* ChangeLog, qore-tibrv-module.spec: added ChangeLog, updated
	  license in spec file to LGPL

2008-09-08 17:49  david_nichols

	* Makefile.am, libtool, qore-tibrv-module.spec: fixed doc packaging

2008-09-06 07:13  david_nichols

	* Makefile.am, qore-tibrv-module.spec: added docs to rpm

2008-09-06 06:57  david_nichols

	* README, docs, docs/docbook, docs/docbook/Makefile,
	  docs/docbook/bookinfo.docbook, docs/docbook/driver.xsl,
	  docs/docbook/module.docbook, docs/docbook/tibrv-module.docbook,
	  docs/img, docs/img/acrobat.gif, docs/img/acrobat_a.gif,
	  docs/img/disk.gif, docs/img/qbullet-note.gif,
	  docs/img/qbullet-note_a.gif, docs/img/remote.gif,
	  docs/img/remote_a.gif, docs/img/scrollup.gif,
	  docs/img/scrollup_a.gif, docs/img/word.gif, docs/qore-style.css,
	  docs/tibrv-module-doc.html: added module documentation

2008-09-03 12:54  david_nichols

	* README: readme update

2008-09-03 12:41  david_nichols

	* src/Makefile.am: build fixes

2008-09-03 12:11  david_nichols

	* COPYING: updated license for modules to LGPL (*not* GPL)

2008-09-03 12:07  david_nichols

	* README: udpated READMEs

2008-09-03 11:57  david_nichols

	* ChangeLog, Makefile.am, configure.ac, libtool,
	  qore-tibrv-module.spec: updated build and fixed spec file

2008-09-03 11:51  david_nichols

	* ., AUTHORS, COPYING, ChangeLog, INSTALL, Makefile.am, NEWS,
	  README, configure.ac, libtool, ltmain.sh, m4, m4/acx_pthread.m4,
	  qore-tibrv-module.spec, reconf.sh, src, src/Makefile.am,
	  src/QC_TibrvCmListener.cc, src/QC_TibrvCmListener.h,
	  src/QC_TibrvCmSender.cc, src/QC_TibrvCmSender.h,

src/tibrv-module.h
/*
  modules/TIBCO/tibrv-module.h

  TIBCO integration to QORE

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef _QORE_TIBRV_MODULE_H

#define _QORE_TIBRV_MODULE_H

#include "../config.h"

#include <qore/Qore.h>

QoreStringNode *tibrv_module_init();
void tibrv_module_ns_init(QoreNamespace *rns, QoreNamespace *qns);
void tibrv_module_delete();

#endif


src/tibrv-module.cc
/*
  modules/TIBCO/tibco-module.cc

  TIBCO integration to QORE

  Qore Programming Language

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "qore_tibrv.h"

#include "QC_TibrvListener.h"
#include "QC_TibrvSender.h"
#include "QC_TibrvFtMember.h"
#include "QC_TibrvFtMonitor.h"
#include "QC_TibrvCmSender.h"
#include "QC_TibrvCmListener.h"
#include "QC_TibrvDistributedQueue.h"
#include "tibrv-module.h"

#include <tibrv/ftcpp.h>

#include <string.h>

DLLEXPORT char qore_module_name[] = "tibrv";
DLLEXPORT char qore_module_version[] = PACKAGE_VERSION;
DLLEXPORT char qore_module_description[] = "TIBCO Rendezvous module";
DLLEXPORT char qore_module_author[] = "David Nichols";
DLLEXPORT char qore_module_url[] = "http://qore.sourceforge.net";
DLLEXPORT int qore_module_api_major = QORE_MODULE_API_MAJOR;
DLLEXPORT int qore_module_api_minor = QORE_MODULE_API_MINOR;
DLLEXPORT qore_module_init_t qore_module_init = tibrv_module_init;
DLLEXPORT qore_module_ns_init_t qore_module_ns_init = tibrv_module_ns_init;
DLLEXPORT qore_module_delete_t qore_module_delete = tibrv_module_delete;
DLLEXPORT qore_license_t qore_module_license = QL_LGPL;

static QoreNamespace *tibns;

static void init_namespace() {
   tibns = new QoreNamespace("Tibrv");
   tibns->addSystemClass(initTibrvListenerClass());
   tibns->addSystemClass(initTibrvSenderClass());
   tibns->addSystemClass(initTibrvFtMemberClass());
   tibns->addSystemClass(initTibrvFtMonitorClass());
   tibns->addSystemClass(initTibrvCmSenderClass());
   tibns->addSystemClass(initTibrvCmListenerClass());
   tibns->addSystemClass(initTibrvDistributedQueueClass());

   // add constants for fault tolerant events
   tibns->addConstant("TIBRVFT_PREPARE_TO_ACTIVATE", new QoreBigIntNode(TIBRVFT_PREPARE_TO_ACTIVATE));
   tibns->addConstant("TIBRVFT_ACTIVATE", new QoreBigIntNode(TIBRVFT_ACTIVATE));
   tibns->addConstant("TIBRVFT_DEACTIVATE", new QoreBigIntNode(TIBRVFT_DEACTIVATE));
   tibns->addConstant("TIBRVFT_QORE_STOP", new QoreBigIntNode(-1));
}

QoreStringNode *tibrv_module_init() {
   // initialize rendezvous
   TibrvStatus status = Tibrv::open();
   if (status != TIBRV_OK) {
      QoreStringNode *err = new QoreStringNode;
      err->sprintf("cannot initialize TIB/RV library: status=%d: %s\n", (int)status, status.getText());
      return err;
   }
   
   init_namespace();

   init_tibrv_functions();
   return NULL;
}

void tibrv_module_ns_init(QoreNamespace *rns, QoreNamespace *qns) {
   QORE_TRACE("tibrv_module_ns_init()");
   qns->addInitialNamespace(tibns->copy());
}

void tibrv_module_delete() {
   QORE_TRACE("tibrv_module_delete()");
   Tibrv::close();
   delete tibns;
}


src/QoreTibrvTransport.cc
/*
  modules/TIBCO/QoreTibrvTransport.cc

  TIBCO Rendezvous integration to QORE

  Qore Programming Language

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include <qore/Qore.h>

#include "QoreTibrvTransport.h"

#include <arpa/inet.h>
#include <stdlib.h>

QoreTibrvTransport::QoreTibrvTransport(const char *desc, const char *service, const char *network, const char *daemon, class ExceptionSink *xsink)
{
   enc = QCS_DEFAULT;
   // create transport (connect to rvd daemon)
   TibrvStatus status = transport.create(service, network, daemon);
   if (status != TIBRV_OK)
   {
      xsink->raiseException("TIBRVTRANSPORT-CONSTRUCTOR-ERROR", "cannot create Rendezvous transport: %s", status.getText());
      return;
   }
   transport.setDescription(desc);
}

int QoreTibrvTransport::hashToMsg(TibrvMsg *msg, const QoreHashNode *hash, ExceptionSink *xsink)
{
   ConstHashIterator hi(hash);
   
   while (hi.next())
   {
      const char *key = hi.getKey();
      if (key && key[0] == '_')
      {
	 xsink->raiseException("TIBRV-MARSHALLING-ERROR", "field names may not begin with '_' (%s)", key);
	 return -1;
      }
      if (valueToField(key, hi.getValue(), msg, xsink))
	 return -1;
   }
   return 0;
}

// add fields to message - we add fields using the name so we can have arrays with different typed elements, etc
// returns 0 for success, -1 for error
int QoreTibrvTransport::valueToField(const char *key, const AbstractQoreNode *v, TibrvMsg *msg, class ExceptionSink *xsink)
{
   //printd(5, "adding %s (%s)\n", key, v ? v->getTypeName() : "null");
   if (is_nothing(v)) {
      msg->addString(key, NULL);
      return 0;
   }

   qore_type_t vtype = v->getType();
   if (vtype == NT_INT) {
      msg->addI64(key, reinterpret_cast<const QoreBigIntNode *>(v)->val);
      return 0;
   }

   if (vtype == NT_FLOAT) {
      msg->addF64(key, reinterpret_cast<const QoreFloatNode *>(v)->f);
      return 0;
   }

   if (vtype == NT_LIST) {
      const QoreListNode *l = reinterpret_cast<const QoreListNode *>(v);
      for (qore_size_t i = 0; i < l->size(); i++)
	 if (valueToField(key, l->retrieve_entry(i), msg, xsink))
	    return -1;
      return 0;
   }

   if (vtype == NT_STRING) {
      const QoreStringNode *str = reinterpret_cast<const QoreStringNode *>(v);
      TempEncodingHelper t(str, enc, xsink);
      if (*xsink)
	 return -1;

      msg->addString(key, t->getBuffer());
      return 0;
   }

   if (vtype == NT_DATE) {
      const DateTimeNode *date = reinterpret_cast<const DateTimeNode *>(v);
      TibrvMsgDateTime dt;      
#ifdef _QORE_HAS_TIME_ZONES
      dt.sec = date->getEpochSecondsUTC();
      dt.nsec = date->getMicrosecond() * 1000;
#else
      dt.sec = date->getEpochSeconds();
      dt.nsec = date->getMillisecond() * 1000000;
#endif
      msg->addDateTime(key, dt);
      return 0;
   }

   if (vtype == NT_HASH) {
      const QoreHashNode *h = reinterpret_cast<const QoreHashNode *>(v);
      //check if it's a type-encoded hash
      const AbstractQoreNode *t;
      if (h->size() == 2 && (t = h->getKeyValue("^type^")) && (t->getType() == NT_STRING))
	 doEncodedType(msg, key, (reinterpret_cast<const QoreStringNode *>(t))->getBuffer(), h->getKeyValue("^value^"), xsink);
      else
      {
	 TibrvMsg m;
	 hashToMsg(&m, h, xsink);
	 if (xsink->isException())
	    return -1;
	 msg->addMsg(key, m);
      }
      return 0;
   }

   if (vtype == NT_BOOLEAN) {
      msg->addBool(key, (tibrv_bool)(reinterpret_cast<const QoreBoolNode *>(v)->getValue()));
      return 0;
   }
   
   if (vtype == NT_BINARY) {
      const BinaryNode *b = reinterpret_cast<const BinaryNode *>(v);
      msg->addOpaque(key, b->getPtr(), b->size());
      return 0;
   }

   xsink->raiseException("TIBRV-MARSHALLING-ERROR", "can't serialize type '%s'", v->getTypeName());
   return -1;
}

QoreHashNode *QoreTibrvTransport::parseMsg(TibrvMsg *msg, class ExceptionSink *xsink)
{
   ReferenceHolder<QoreHashNode> data(msgToHash(msg, xsink), xsink);
   if (xsink->isException())
      return 0;
   
   QoreHashNode *h = new QoreHashNode();
   h->setKeyValue("msg", data.release(), NULL);
   
   const char *str;
   TibrvStatus status = msg->getReplySubject(str);
   if (status == TIBRV_OK)
      h->setKeyValue("replySubject", new QoreStringNode(str), xsink);
   
   status = msg->getSendSubject(str);
   if (status == TIBRV_OK)
      h->setKeyValue("subject", new QoreStringNode(str), xsink);

   return h;
}

QoreHashNode *QoreTibrvTransport::msgToHash(TibrvMsg *msg, class ExceptionSink *xsink)
{
   tibrv_u32 len;
   TibrvStatus status = msg->getNumFields(len);
   if (status != TIBRV_OK)
   {
      xsink->raiseException("TIBRV-DEMARSHALLING-ERROR", status.getText());
      return 0;
   }

   TibrvMsgField field;

   ReferenceHolder<QoreHashNode> h(new QoreHashNode(), xsink);
   for (unsigned i = 0; i < len; i++)
   {
      status = msg->getFieldByIndex(field, i);
      if (status != TIBRV_OK)
      {
	 xsink->raiseException("TIBRV-DEMARSHALLING-ERROR", status.getText());
	 return 0;
      }

      const char *key = field.name;
      // if a null pointer is found, then change to a zero-length string
      if (!key)
	 key = "";

      AbstractQoreNode *val = fieldToNode(&field, xsink);
      if (xsink->isException())
	 return 0;

      bool exists;
      AbstractQoreNode *ev = h->getKeyValueExistence(key, exists);
      if (exists)
      {
	 AbstractQoreNode **evp = h->getKeyValuePtr(key);
	 QoreListNode *l = dynamic_cast<QoreListNode *>(ev);
	 if (!l) {
	    //printd(5, "QoreTibrvTransport::msgToHash() making list\n");
	    l = new QoreListNode();
	    l->push(ev);
	    ev = l;
	 }
	 l->push(val);
	 *evp = ev;
      }
      else
	 h->setKeyValue(key, val, NULL);
   }

   return h.release();
}

AbstractQoreNode *QoreTibrvTransport::fieldToNode(TibrvMsgField *field, class ExceptionSink *xsink)
{
   if (field->count > 1)
      return listToNode(field, xsink);

   tibrvLocalData data = field->getData();

   //printd(5, "QoreTibrvTransport::fieldToNode() type=%d\n", field->getType());

   switch (field->getType())
   {
      case TIBRVMSG_STRING:
	 return new QoreStringNode((char *)data.str, enc);

      case TIBRVMSG_I8:
	 return new QoreBigIntNode(data.i8);

      case TIBRVMSG_U8:
	 return new QoreBigIntNode(data.u8);

      case TIBRVMSG_I16:
	 return new QoreBigIntNode(data.i16);

      case TIBRVMSG_U16:
	 return new QoreBigIntNode(data.u16);

      case TIBRVMSG_I32:
	 return new QoreBigIntNode(data.i32);

      case TIBRVMSG_U32:
	 return new QoreBigIntNode(data.u32);

      case TIBRVMSG_I64:
	 return new QoreBigIntNode(data.i64);

      case TIBRVMSG_U64: // WARNING: we lose precision here!!!
	 return new QoreBigIntNode((int64)data.u64);

      case TIBRVMSG_F32:
	 return new QoreFloatNode((double)data.f32);
 
      case TIBRVMSG_F64:
	 return new QoreFloatNode(data.f64);

      case TIBRVMSG_IPPORT16:
	 return new QoreBigIntNode(ntohs(data.ipport16));

      case TIBRVMSG_IPADDR32:
      {
	 class QoreStringNode *addr = new QoreStringNode(enc);
	 unsigned int ia = ntohl(data.ipaddr32);
	 //printd(0, "ipaddr32: data=%u conv=%u\n", data.ipaddr32, ia);
	 unsigned char *uc = (unsigned char*)&ia;
	 addr->sprintf("%d.%d.%d.%d", (int)uc[3], (int)uc[2], (int)uc[1], (int)uc[0]);
	 //printd(5, "addr=%s\n", addr->getBuffer());
	 return addr;
      }

      case TIBRVMSG_BOOL:
	 return get_bool_node((bool)data.boolean);

      case TIBRVMSG_MSG:
      {
	 //printd(5, "TIBRVMSG_MSG size=%d\n", field->size);
	 TibrvMsg msg(data.msg, TIBRV_FALSE);
	 return msgToHash(&msg, xsink);
      }

      case TIBRVMSG_DATETIME:
#ifdef _QORE_HAS_TIME_ZONES
	 return DateTimeNode::makeAbsolute(currentTZ(), (int64)data.date.sec, data.date.nsec / 1000);
#else
	 return new DateTimeNode((int64)data.date.sec, data.data.nsec / 1000000);
#endif

      case TIBRVMSG_OPAQUE:
      {
	 void *ptr = malloc(field->size);
	 memcpy(ptr, data.buf, field->size);
	 return new BinaryNode(ptr, field->size);
      }

      case TIBRVMSG_XML:
      {
	 class QoreStringNode *str = new QoreStringNode(enc);
	 // ensure that xml string is terminated with a null character in case it wasn't sent
	 int len = field->size;
	 if (!((char *)data.buf)[len - 1])
	    len--;
	 str->terminate(len);
	 memcpy((char *)str->getBuffer(), data.buf, len);
	 return str;
      }
   }

   xsink->raiseException("TIBRV-DEMARSHALLING-ERROR", "don't know how to convert type %d", field->getType());
   return 0;
}

class AbstractQoreNode *QoreTibrvTransport::listToNode(TibrvMsgField *field, class ExceptionSink *xsink)
{
   ReferenceHolder<QoreListNode> l(new QoreListNode(), xsink);
   tibrvLocalData data = field->getData();

   switch (field->getType())
   {
      case TIBRVMSG_I8:
      {
	 char *c = (char *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreBigIntNode(c[i]));
	 break;
      }

      case TIBRVMSG_U8:
      {
	 unsigned char *c = (unsigned char *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreBigIntNode(c[i]));
	 break;
      }
      case TIBRVMSG_I16:
      {
	 short *c = (short *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreBigIntNode(c[i]));
	 break;
      }

      case TIBRVMSG_U16:
      {
	 unsigned short *c = (unsigned short *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreBigIntNode(c[i]));
	 break;
      }

      case TIBRVMSG_I32:
      {
	 int *c = (int *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreBigIntNode(c[i]));
	 break;
      }

      case TIBRVMSG_U32:
      {
	 unsigned *c = (unsigned *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreBigIntNode(c[i]));
	 break;
      }

      case TIBRVMSG_U64:
      case TIBRVMSG_I64:
      {
	 int64 *c = (int64 *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreBigIntNode(c[i]));
	 break;
      }

      case TIBRVMSG_F32:
      {
	 float *c = (float *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreFloatNode((double)c[i]));
	 break;
      }

      case TIBRVMSG_F64:
      {
	 double *c = (double *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(new QoreFloatNode(c[i]));
	 break;
      }

      case TIBRVMSG_BOOL:
      {
	 tibrv_bool *c = (tibrv_bool *)data.buf;
	 for (unsigned i = 0; i < field->count; i++)
	    l->push(get_bool_node((bool)c[i]));
	 break;
      }

      default:
	 xsink->raiseException("TIBRV-DEMARSHALLING-ERROR", "don't know how to convert a list of type %d", field->getType());
	 return 0;
   }
   return l.release();
}

int QoreTibrvTransport::doEncodedType(TibrvMsg *msg, const char *key, const char *type, const AbstractQoreNode *val, class ExceptionSink *xsink)
{
   if (type[0] == 'i')
   {
      type++;
      if (!strcmp(type, "8"))
      {
	 msg->addI8(key, val ? val->getAsInt() : 0);
	 return 0;
      }
      if (!strcmp(type, "16"))
      {
	 msg->addI16(key, val ? val->getAsInt() : 0);
	 return 0;
      }
      if (!strcmp(type, "32"))
      {
	 msg->addI32(key, val ? val->getAsInt() : 0);
	 return 0;
      }
      if (!strcmp(type, "64"))
      {
	 msg->addI64(key, val ? val->getAsBigInt() : 0);
	 return 0;
      }
      else if (!strcmp(type, "pport16"))
      {
	 msg->addIPPort16(key, htons(val ? val->getAsInt() : 0));
	 return 0;
      }
      else if (!strcmp(type, "paddr32"))
      {
	 const QoreStringNode *vstr = dynamic_cast<const QoreStringNode *>(val);
	 if (!vstr) {
	    xsink->raiseException("TIBRV-MARSHALLING-ERROR", "can't serialize tibrv type 'ipaddr32' from qore type '%s' (expecting string)", val ? val->getTypeName() : "NOTHING");
	    return -1;
	 }

	 // assuming format "#.#.#.#"
	 int addr[4] = { 0, 0, 0, 0 };
	 int i = 0;
	 const char *c;
	 const char *buf = vstr->getBuffer();
	 QoreString str;
	 while ((c = strchr(buf, '.')))
	 {
	    str.terminate(0);
	    str.concat(buf, c - buf);
	    addr[i] = atoi(str.getBuffer());
	    //printd(5, "addr[%d]=%d\n", i, addr[i]);
	    i++;
	    if (i == 3)
	    {
	       addr[i] = atoi(c + 1);
	       //printd(5, "addr[%d]=%d\n", i, addr[i]);
	       break;
	    }
	    buf = c + 1;
	 }
	 // put it network byte order directly
	 i = addr[0] | addr[1] << 8 | addr[2] << 16 | addr[3] << 24;
	 msg->addIPAddr32(key, i);
	 
	 return 0;
      }
   }
   else if (type[0] == 'u')
   {   
      type++;
      if (!strcmp(type, "8"))
      {
	 msg->addU8(key, val ? val->getAsInt() : 0);
	 return 0;
      }
      
      if (!strcmp(type, "16"))
      {
	 msg->addU16(key, val ? val->getAsInt() : 0);
	 return 0;
      }
      
      if (!strcmp(type, "32"))
      {
	 msg->addU32(key, val ? val->getAsInt() : 0);
	 return 0;
      }
      
      if (!strcmp(type, "64"))
      {
	 msg->addU64(key, val ? val->getAsBigInt() : 0);
	 return 0;
      }
   }
   else if (type[0] == 'f')
   {
      type++;
      if (!strcmp(type, "32"))
      {
	 msg->addF32(key, val ? val->getAsFloat() : 0.0);
	 return 0;
      }
      
      if (!strcmp(type, "64"))
      {
	 msg->addF64(key, val ? val->getAsFloat() : 0.0);
	 return 0;
      }
   }
   else if (!strcmp(type, "bool"))
   {
      msg->addBool(key, (tibrv_bool)(val ? val->getAsBool() : false));
      return 0;
   }
   else if (!strcmp(type, "xml"))
   {
      const QoreStringNode *str = dynamic_cast<const QoreStringNode *>(val);
      if (!str)
      {
	 xsink->raiseException("TIBRV-MARSHALLING-ERROR", "can't serialize tibrv type 'xml' from qore type '%s' (need string)", val ? val->getTypeName() : "NOTHING");
	 return -1;
      }

      msg->addXml(key, str->getBuffer(), str->strlen() + 1);
      return 0;
   }

   xsink->raiseException("TIBRV-MARSHALLING-ERROR", "don't know how to serialize tibrv type '%s'", type);
   return -1;
}
